core/field_variable.js

```diff
@@ -159,39 +168,50 @@ Blockly.FieldVariable.dropdownCreate = function() {
  * @this {!Blockly.FieldVariable}
  */
 Blockly.FieldVariable.dropdownChange = function(text) {
-  function promptName(promptText, defaultText) {
+  var oldVar = this.getText();
+  var workspace = this.sourceBlock_.workspace;
+  var dialog = workspace.options.hasModal;
+  var callbackFunc = function(variable_name) {
     Blockly.hideChaff();
-    var newVar = window.prompt(promptText, defaultText);
+    if (!variable_name) {
+      return;
+    }
     // Merge runs of whitespace.  Strip leading and trailing whitespace.
     // Beyond this, all names are legal.
-    if (newVar) {
-      newVar = newVar.replace(/[\s\xa0]+/g, ' ').replace(/^ | $/g, '');
-      if (newVar == Blockly.Msg.RENAME_VARIABLE ||
-          newVar == Blockly.Msg.NEW_VARIABLE) {
-        // Ok, not ALL names are legal...
-        newVar = null;
-      }
+    variable_name = variable_name
+      .replace(/[\s\xa0]+/g, ' ')
+      .replace(/^ | $/g, '');
+    if (!variable_name ||
+        variable_name == Blockly.Msg.RENAME_VARIABLE ||
+        variable_name == Blockly.Msg.NEW_VARIABLE) {
+      // Ok, not ALL names are legal...
+      return;
+    }
+
+    if (text == Blockly.Msg.RENAME_VARIABLE) {
+      Blockly.Variables.renameVariable(oldVar, variable_name, workspace);
+    } else if (text == Blockly.Msg.NEW_VARIABLE) {
+      Blockly.Variables.renameVariable(text.replace('...', ''), variable_name,
+        workspace);
     }
-    return newVar;
+  };
+
+  if (!dialog) {
+    dialog = function(promptText, defaultText, callback) {
+      var variableName = window.prompt(promptText, defaultText);
+      callback(variableName);
+    };
   }
-  var workspace = this.sourceBlock_.workspace;
+
   if (text == Blockly.Msg.RENAME_VARIABLE) {
-    var oldVar = this.getText();
-    text = promptName(Blockly.Msg.RENAME_VARIABLE_TITLE.replace('%1', oldVar),
-                      oldVar);
-    if (text) {
-      Blockly.Variables.renameVariable(oldVar, text, workspace);
-    }
+    dialog(Blockly.Msg.RENAME_VARIABLE_TITLE.replace('%1', oldVar), oldVar,
+        callbackFunc);
     return null;
   } else if (text == Blockly.Msg.NEW_VARIABLE) {
-    text = promptName(Blockly.Msg.NEW_VARIABLE_TITLE, '');
-    // Since variables are case-insensitive, ensure that if the new variable
-    // matches with an existing variable, the new case prevails throughout.
-    if (text) {
-      Blockly.Variables.renameVariable(text, text, workspace);
-      return text;
-    }
-    return null;
+    window.setTimeout(function() {
+      dialog(Blockly.Msg.NEW_VARIABLE_TITLE, '', callbackFunc);
+    }, 100);
+    return Blockly.Msg.NEW_VARIABLE.replace('...', '');
```

Comment:
This could clash with a variable named "New Variable", which is not a very far fetched name. You could instead generate a unique name and use that instead, I can't remember from the top of my head what the function was, but it is probably defined in the variables.js file.